# Homework 5

## Question 1 

Consider the function $f(x)=\sin{x}$.

(a) Compute the *forward difference* approximation of $f'(0.5)$ using $h=0.1$.

Recall, the *forward difference* approximation is given by, 
$$f'(x_0)=\frac{f(x_0+h)-f(x_0)}{h}-\frac{h}{2}f''(\xi).$$

In [2]:
import numpy as np
import math

In [3]:
# implement forward difference formula 
def forwardDifference(f,x0,h):
    """
    forwardDifference - computes the forward difference given by
    the forward difference formula for numerical approximation
    of a function's derivative; takes a function object, 
    evaluation point, and value for h in the difference
    quotient.
    """
    return (f(x0+h)-f(x0))/h

In [7]:
# implement given function 
def f(x):
    return math.sin(x)

In [8]:
# compute forward difference approximation at x0=0.5, h=0.1
forwardDifference(f,0.5,0.1)

0.8521693479083237

Thus the forward approximation $\approx 0.8521693479083237$.

(b) Compute the *backward difference* approximation of $f'(0.7)$ using $h=-0.1$. 

Recall, the *backward difference* approximation is given by,
$$f'(x_0)=\frac{f(x_0)-f(x_0-h)}{h}+\frac{h}{2}f''(\xi).$$

In [9]:
# implement backward difference formula
def backwardDifference(f,x0,h):
    """
    backwardDifference - computes the backward difference given by
    the backward difference formula for numerical approximation
    of a function's derivative; takes a function object, 
    evaluation point, and value for h in the difference
    quotient.
    """
    return (f(x0)-f(x0-h))/h

In [10]:
# compute backward difference approximation at x0=0.7, h=0.1
# (h=-0.1 if using the forward difference version)
backwardDifference(f,0.7,0.1)

0.7957521384265565

Thus the backward approximation $\approx 0.7957521384265565$.

---

## Question 2

Compute the actual error produced by the approximations in problem 1, and find the error bounds using the error formula.

To compute the actual error, we need the exact value of the derivative at $0.5$ and $0.7$, then find the difference from our approximated values.
The first derivative of $f(x)=\sin{x}$ is $f'(x)=\cos{x}$. 

In [32]:
# implement first derivative (cos(x))
def fp(x):
    """
    Simply evaluates cosine function at x
    """
    return math.cos(x)

In [33]:
# vectorize first derivative function to allow for vector arguments
fpv = np.vectorize(fp)

In [34]:
# compute exact values 
vals = np.array([0.5,0.7])
fpv(vals)

array([0.87758256, 0.76484219])

Hence our exact value for $f'(0.5)=0.87758256$ and the exact value for $f'(0.7)=0.76484219$

In [35]:
# implement error function to compute both at same time 
def error(exact,approx):
    """
    Simply computes the actual error using a numeric vector 
    of approximated values and a numeric vector of exact values
    """
    return math.fabs(exact-approx)

In [36]:
# compute actual error for x0=0.5 (used forward error to approximate)
error(fp(0.5),forwardDifference(f,0.5,0.1))

0.0254132139820491

Thus the actual error for our forward difference approximation of $f'(0.5)$ using $h=0.1$ is $\approx 0.0254132139820491$.

In [37]:
# compute actual error for x0=0.7 (used backward error to approximate)
error(fp(0.7),backwardDifference(f,0.7,0.1))

0.03090995114206796

Thus the actual error for our backward difference approximation is $\approx 0.03090995114206796.$

Note that for both of these values when computed by hand and using four decimals differ slightly at about $0.0255$ and $0.0312$. 
These slight differences are possibly due to some rounding or truncation error in conjunction with having to divide by small values in the difference quotient. 

To compute the error bounds for these approximations, we'll use the last term in both the forward and back difference formulas. The error is bounded by $\frac{M|h|}{2}$ where $M$ is a bound on $|f''(x)|$ for $x_0<x<x_0+h$ for the forward difference and $x_0-h<x<x_0$ for the backward difference. 

For this particular problem, our second derivative is given by,
$$f''(x)=-\sin{x}.$$

Then for the approximation of $f'(0.5)$ using $h=0.1$, the error bound is estimated by, 
$$ \Rightarrow \frac{|hf''(\xi)|}{2} = \frac{|h|}{2}|f''(\xi)| = \frac{0.1}{2}\left|\sup \limits_{0.5<\xi<0.6}(-\sin{\xi})\right| = \frac{0.1}{2}\left|-\sin{0.6}\right| 
\approx 0.02823.$$

For the approximation of $f'(0.7)$ using $h=0.1$, the error bound is estimated by, 
$$ \Rightarrow \frac{|hf''(\xi)|}{2} = \frac{|h|}{2}|f''(\xi)| = \frac{0.1}{2}\left|\sup \limits_{0.6<\xi<0.7}(-\sin{\xi})\right| = \frac{0.1}{2}\left|-\sin{0.7}\right| 
\approx 0.03221.$$